keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+
+With the new methods we would like also to migrate to klog v2 and introduce a new JSON logging output format to Kubernetes components that would be an alternative to the current text format and which will make querying and processing Kubernetes logs even simpler.
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
+
+We would like for the Kubernetes community to settle on one preferred log message structure. Proposed structure should:
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using “%q” fmt logic
+* keys are formatted using “%s” fmt logic
+* values are formatted using “%v” fmt logic and optionally escaped using the “%q” fmt logic if it not a number or a boolean
+* keys are separated from values with a “=” sign
+* message and key-values are separated with a single space
+
+#### New klog methods
+
+To enforce a new log message structure we want to introduce new methods to klog library which will provide a more structured interface for formatting log messages compared to current methods based on fmt format strings.
+For each format method (`Infof`, `Warningf`, `Errorf`, `Fatalf`) we will add matching structured method (`InfoS`, `WarningS`, `ErrorS`, `FatalS`).
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] “Pod status updated“ pod=“kubedns“ status=“ready“
+```
+
+### References to Kubernetes objects
+
+Currently there is no consistency when logging references to Kubenetes objects like Pods. Depending on a log just the name of a Pod, the name and namespace as separate fields or Pod name prefixed with a namespace and ‘/’ character is logged.
+This makes finding logs related to a given object much harder.
+To address this problem we want to implement helper functions for all standard object types which could be used to build values which when logged will provide a consistent identification of each of those objects.
+For instance for namespaced objects those methods should produce following value: “<namespace>/<name>”.
+
+Example:
+```go
+klog.InfoS("Pod status updated", "pod", metadata.Pod(pod), "status", "ready")
```

Comment:
I was under the assumption that last we spoke about this, any of these helper functions would be out of scope for the first iteration. Has this changed?